scene/2d/node_2d.cpp

```diff
@@ -385,9 +385,14 @@ Point2 Node2D::to_global(Point2 p_local) const {
 
 void Node2D::_notification(int p_notification) {
 	switch (p_notification) {
-		case NOTIFICATION_MOVED_IN_PARENT: {
+		case NOTIFICATION_ENTER_TREE: {
 			if (get_viewport()) {
-				get_viewport()->gui_set_root_order_dirty();
+				get_parent()->connect(SNAME("children_list_changed"), callable_mp(get_viewport(), &Viewport::gui_set_root_order_dirty), CONNECT_REFERENCE_COUNTED);
```

Comment:
Shouldn't this be connected in CanvasItem instead? It's the same as in Control.

Reply:
No idea, I think the logic is different for controls, but this is how the original code works, so I don't want to change it in this PR.